pkg/apis/tenancy/v1alpha1/types_workspacetype.go

```diff
@@ -0,0 +1,267 @@
+	//
+	// +optional
+	Extend WorkspaceTypesExtension `json:"extend,omitempty"`
+
+	// additionalWorkspaceLabels are a set of labels that will be added to a
+	// ClusterWorkspace on creation.
+	//
+	// +optional
+	AdditionalWorkspaceLabels map[string]string `json:"additionalWorkspaceLabels,omitempty"`
+
+	// defaultChildWorkspaceType is the WorkspaceType that will be used
+	// by default if another, nested ClusterWorkspace is created in a workspace
+	// of this type. When this field is unset, the user must specify a type when
+	// creating nested workspaces. Extending another WorkspaceType does
+	// not inherit its defaultChildWorkspaceType.
+	//
+	// +optional
+	DefaultChildWorkspaceType *WorkspaceTypesReference `json:"defaultChildWorkspaceType,omitempty"`
+
+	// limitAllowedChildren specifies constraints for sub-workspaces created in workspaces
+	// of this type. These are in addition to child constraints of types this one extends.
+	//
+	// +optional
+	LimitAllowedChildren *WorkspaceTypeSelector `json:"limitAllowedChildren,omitempty"`
+
+	// limitAllowedParents specifies constraints for the parent workspace that workspaces
+	// of this type are created in. These are in addition to parent constraints of types this one
+	// extends.
+	//
+	// +optional
+	LimitAllowedParents *WorkspaceTypeSelector `json:"limitAllowedParents,omitempty"`
+
+	// defaultAPIBindings are the APIs to bind during initialization of workspaces created from this type.
+	// The APIBinding names will be generated dynamically.
+	//
+	// +optional
+	DefaultAPIBindings []APIExportReference `json:"defaultAPIBindings,omitempty"`
+}
+
+// APIExportReference provides the fields necessary to resolve an APIExport.
+type APIExportReference struct {
+	// path is the fully-qualified path to the workspace containing the APIExport. If it is
+	// empty, the current workspace is assumed.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern:="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(:[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+	Path string `json:"path,omitempty"`
+
+	// export is the name of the APIExport.
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	// +kube:validation:MinLength=1
+	Export string `json:"export"`
+}
+
+// WorkspaceTypeSelector describes a set of types.
+type WorkspaceTypeSelector struct {
+	// none means that no type matches.
+	//
+	// +kuberbuilders:Enum=true
+	None bool `json:"none,omitempty"`
+
+	// types is a list of WorkspaceTypes that match. A workspace type extending
+	// another workspace type automatically is considered as that extended type as well
+	// (even transitively).
+	//
+	// An empty list matches all types.
+	//
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	Types []WorkspaceTypesReference `json:"types,omitempty"`
+}
+
+// WorkspaceTypesExtension defines how other WorkspaceTypes are
+// composed together to add functionality to the owning WorkspaceType.
+type WorkspaceTypesExtension struct {
```

Comment:
Global comment: most `WorkspaceTypes<text>` instances should be WorkspaceType, singular?

Reply:
done